src/AbstractApi.ts

```diff
@@ -1,68 +1,103 @@
-import IApi, { FetchMethod } from './IApi';
-import { HttpMethods, AuthSchemes } from './Enums';
+import {AuthSchemes, HttpMethods} from './Enums';
 import AuthService from './AuthService';
+import ILoadingProvider from './ILoadingProvider';
+import IInitOptions from './IInitOptions';
 
 export default abstract class AbstractApi {
   readonly apiRoot: string;
   readonly apiRequiresAuth: boolean;
   hasAuthService: boolean = false;
-  private authService?: AuthService;
+  private readonly authService?: AuthService;
+  protected readonly loadingProvider?: ILoadingProvider;
+
+  constructor(apiRoot: string, requiresAuth: boolean, initOptions: IInitOptions = {}) {
+    const {
+      getToken,
+      authScheme,
+      loadingProvider
+    } = initOptions;
 
-  constructor(apiRoot: string, requiresAuth: boolean, getToken?: (...args: any[]) => string, authScheme: AuthSchemes = AuthSchemes.Bearer) {
     this.apiRoot = apiRoot;
     this.apiRequiresAuth = requiresAuth;
     if (requiresAuth) {
       if (!getToken) {
         throw new Error('Must pass token retrieval logic as a function if requiresAuth == true.');
       }
-      this.authService = new AuthService(authScheme, getToken);
+      this.authService = new AuthService(authScheme || AuthSchemes.Bearer, getToken);
       this.hasAuthService = true;
     }
+
+    if (!!loadingProvider) {
+      if (!loadingProvider.onBegin || !loadingProvider.onResolve) {
+        throw new Error('Must provide an `onBegin` and and `onResolve` function to loadingProvider');
+      }
+      this.loadingProvider = loadingProvider;
+    }
   }
 
-  protected abstract async resolve(response: Response): Promise<any>;
+  protected abstract async resolve(response: Response, handleLoading: boolean): Promise<any>;
+
+  async fetch(url: string, options: IRequestOptions): Promise<any> {
+    // we should handle loading by default if there's a loading provider and no `handleLoading`
+    // key in our options object. Otherwise, if there is a loading provider, we go by the value of the
+    // handleLoading option. If no loadingProvider, we will never handle loading.
+    let handleLoading = false;
+    if (!!this.loadingProvider) {
+      handleLoading = !(options.hasOwnProperty('handleLoading') && options.handleLoading === false);
```

Comment:
I would suggest default to have false and not true

Reply:
The problem is that doing so would break what's currently in hf-client. We'd have to run through every single API call and add this option.